checkout-example-advanced/index.js

```diff
@@ -0,0 +1,287 @@
+config.apiKey = process.env.ADYEN_API_KEY;
+const client = new Client({ config });
+client.setEnvironment("TEST");
+const checkout = new CheckoutAPI(client);
+
+app.engine(
+  "handlebars",
+  hbs({
+    defaultLayout: "main",
+    layoutsDir: __dirname + "/views/layouts",
+    helpers: require("./util/helpers"),
+  })
+);
+
+app.set("view engine", "handlebars");
+
+/* ################# API ENDPOINTS ###################### */
+
+// Get payment methods
+app.post("/api/getPaymentMethods", async (req, res) => {
+  try {
+    const response = await checkout.paymentMethods({
+      channel: "Web",
+      merchantAccount: process.env.ADYEN_MERCHANT_ACCOUNT,
+    });
+    res.json(response);
+  } catch (err) {
+    console.error(`Error: ${err.message}, error code: ${err.errorCode}`);
+    res.status(err.statusCode).json(err.message);
+  }
+});
+
+// Submitting a payment
+app.post("/api/initiatePayment", async (req, res) => {
+  const currency = findCurrency(req.body.paymentMethod.type);
+  // find shopper IP from request
+  const shopperIP = req.headers["x-forwarded-for"] || req.connection.remoteAddress;
+
+  try {
+    // unique ref for the transaction
+    const orderRef = uuid();
+    // Ideally the data passed here should be computed based on business logic
+    const response = await checkout.payments({
+      amount: { currency, value: 1000 }, // value is 10€ in minor units
+      reference: orderRef, // required
+      merchantAccount: process.env.ADYEN_MERCHANT_ACCOUNT, // required
+      channel: "Web", // required
+      additionalData: {
+        // required for 3ds2 native flow
+        allow3DS2: true,
+      },
+      origin: "http://localhost:8080", // required for 3ds2 native flow
+      browserInfo: req.body.browserInfo, // required for 3ds2
+      shopperIP, // required by some issuers for 3ds2
+      returnUrl: `http://localhost:8080/api/handleShopperRedirect?orderRef=${orderRef}`, // required for 3ds2 redirect flow
+      // special handling for boleto
+      paymentMethod: req.body.paymentMethod.type.includes("boleto")
+        ? {
+            type: "boletobancario_santander",
+          }
+        : req.body.paymentMethod,
+      // Below fields are required for Boleto:
+      socialSecurityNumber: req.body.socialSecurityNumber,
+      shopperName: req.body.shopperName,
+      billingAddress:
+        typeof req.body.billingAddress === "undefined" || Object.keys(req.body.billingAddress).length === 0
+          ? null
+          : req.body.billingAddress,
+      deliveryDate: "2023-12-31T23:00:00.000Z",
+      shopperStatement: "Aceitar o pagamento até 15 dias após o vencimento.Não cobrar juros. Não aceitar o pagamento com cheque",
+      // Below fields are required for Klarna:
+      countryCode: req.body.paymentMethod.type.includes("klarna") ? "DE" : null,
+      shopperReference: "12345",
+      shopperEmail: "youremail@email.com",
+      shopperLocale: "en_US",
+      lineItems: [
+        {
+          quantity: "1",
```

Comment:
Sunglasses / Headphones, 50.00 + 50.00 = 100.00

lineItems do not have the correct values for amount, quantities, taxPercantages etc.